tests/integration/test_id_match.py

```diff
@@ -0,0 +1,214 @@
+    id_ = flow({
+        'scheduling': {
+            'cycling mode': 'integer',
+            'initial cycle point': '1',
+            'final cycle point': '3',
+            'graph': {
+                'P2': '''
+                    a1 => b1 => c1
+                    a2 => b2 => c2
+
+                    b1[-P1] => b1
+                    b2[-P1] => b2
+                ''',
+            },
+        },
+        'runtime': {
+            'a1, a2': {'inherit': 'A'},
+            'A': {},
+            'b1, b2': {'inherit': 'B'},
+            'B': {},
+            'c1, c2': {},
+        },
+    })
+    schd: Scheduler = scheduler(id_)
+
+    def match(*ids: str) -> Tuple[Set[str], Set[str]]:
+        matched, unmatched = schd.pool.id_match(
+            {cast('TaskTokens', Tokens(id_, relative=True)) for id_ in ids},
+        )
+        return {id_.relative_id for id_ in matched}, {
+            id_.relative_id_with_selectors for id_ in unmatched
+        }
+
+    async with start(schd):
+        await commands.run_cmd(
+            commands.set_prereqs_and_outputs(
+                schd, ['1/a2'], ['1'], ['succeeded'], None
+            )
+        )
+        await commands.run_cmd(
+            commands.set_prereqs_and_outputs(
+                schd, ['1/b2'], ['1'], ['failed'], None
+            )
+        )
+
+        # task pool state:
+        # * cycle 1
+        #   * n=0 a1 waiting
+        #   * n=1 b1 waiting
+        #   * n=2 c1 waiting
+        #   * n=1 a2 succeeded
+        #   * n=0 b2 failed
+        #   * n=1 c2 waiting
+        # * cycle 2
+        #   * n=0 a1 waiting
+        #   * n=1 b1 waiting
+        #   * n=2 c1 waiting
+        #   * n=0 a2 waiting
+        #   * n=1 b2 waiting
+        #   * n=2 c2 waiting
+
+        # check the n=0 window matches expecations before proceeding
+        assert {
+            itask.tokens.relative_id for itask in schd.pool.get_tasks()
+        } == {'1/a1', '1/b2', '3/a1', '3/a2'}
+
+        # test active task selection
+        assert (
+            match('*:waiting')
+            == match('*/root:waiting')
+            == match('*/*:waiting')
+            == match('*/A:waiting')
+            == match('*/a*:waiting')
+            == match('1/a1:waiting', '3/a1:waiting', '3/a2:waiting')
+            == match('^/a1:waiting', '3/a1:waiting', '$/a2:waiting')
+            == ({'1/a1', '3/a1', '3/a2'}, set())
+        )
+
+        assert (
+            match('*:failed')
+            == match('*/root:failed')
+            == match('*/*:failed')
+            == match('*/B:failed')
+            == match('*/b*:failed')
+            == match('1/b2:failed')
+            == match('^/b2:failed')
+            == ({'1/b2'}, set())
+        )
+
+        assert (
+            match('1/b1:failed', '1/b2:failed')
+            == match('1/B:failed', '1/b1:failed')
+            == match('*:failed', '1/B:failed', '1/b1:failed')
+            == ({'1/b2'}, {'1/b1:failed'})
```

Comment:
The unmatched result here surprised me. I'll be poking this to see if I'm unhappy with the consequences.